Add new holland mariadb-dump plugin and various bug fixes

ed-velez opened the pull request:
Adds a new mariadb-dump plugin to facilitate the diversion of mariabd-dump from mysqldump occurring upstream. This started off largely as a copy and paste of the mysqldump plugin and some search and replace fun. It does have some diversions from mysqldump that I'm open to changing.

Changes from mysqldump:
- The backupset provider conf for this plugin includes all options we expose via the configspec.
- The configspec for mariadb-dump uses a generic `executable` setting that defaults to `mariadb-dump` and uses a new `get_cmd_path` method helper that ensure we properly search the holland defined environment PATH. This value can also be set to an absolute path. This also allows for the executable to be named anything so that a user could potentially even wrap the mariadb-dump call if needed for some reason. The existing `which` helper has some inconsistent behavior differences when not using `shutil.which` that should be addressed with the new helper. Plugins should migrate to using this new helper.
- The `command.py` in mysqldump was overly complicated and didn't provide us much benefit. I've simplified that interface a bit so that the validation made more sense and was straight forward. Any option not exposed via the configspec and is not something that we are providing some validation is not anything we need to care about in code: https://github.com/holland-backup/holland/blob/master/plugins/holland.backup.mysqldump/holland/backup/mysqldump/command.py#L100-L136. 
- I've removed the `info` helper for this plugin when running `holland list-backups -v`. Not all plugins provide it and the existing verbosity seems sufficient.

General Bug fixes include:

- `holland list-backups -v` does not properly template the values in it's output due to the hyphens used in the template string. This has been fixed.
- `holland list-backups -v` also complained about `failed` being an unknown option it found in the config read from the backup. We defined `failed-backup` in the core configspec - but were maintaining state with the value `failed`. I changed `failed-backup` to `failed` to match what we were actually using in code.
- The existing mysqldump plugin has this code path of only adding `flush-logs` to the last database in the list under certain conditions:
https://github.com/holland-backup/holland/blob/master/plugins/holland.backup.mysqldump/holland/backup/mysqldump/base.py#L46-L50
However, this condition will never be hit as enumerate starts at 0 and checking the list length for equality will never be true. I fixed the logic in this mariadb-dump plugin but need to confirm its a bug since since no one has mentioned anything about it.

Setting this to draft for now as I continue to review these changes.

mikegriffin commented:
I like it, with some minor feedback:

- [ ] bikeshedding - maybe package name holland.backup.mariadb_dump => holland.backup.mariadb-dump to match binary name like pg_basebackup? plugin name is already mariadb-dump so this isn't a big deal

- [ ]  manual install instructions in INSTALL.md mention python instead of python3; also requires plugins/holland.lib.common/ probably in most cases, perhaps this even replaces recommendation of holland.lib.mysql?

- [ ]  minor make-config default change for this new provider
```
estimated-size-factor = 1.0
estimated-size-factor = 0.20
```

- [ ]  MariaDB uses PAM auth by default but provider doesn't default to using unix socket but instead uses localhost, I think a standard unix socket path is a good clue to new users rather than intuition of setting a password for existing root db user. Proposed fix, make-config:
```
### WAS
[mysql:client]
defaults-extra-file = ~/.my.cnf,
# user = "" # no default
# password = "" # no default
# socket = "" # no default
# host = "" # no default
# port = "" # no default
```
```
### NEW
[mysql:client]
defaults-extra-file = ~/.my.cnf,
user = "root"
# password = "" # no default, modern mariadb uses pam auth
socket = "/var/lib/mysql/mysql.sock" # no default
# host = "" # no default
# port = "" # no default
```

ed-velez commented:
> I like it, with some minor feedback:
> 
>     * [ ]  bikeshedding - maybe package name holland.backup.mariadb_dump => holland.backup.mariadb-dump to match binary name like pg_basebackup? plugin name is already mariadb-dump so this isn't a big deal
> 

Hyphens in python package names are generally avoided. This is similar to why the python module for mysql-lvm is `holland.backup.mysql_lvm` but the plugin name is `mysql-lvm`. But yeah I tried to at the very least match the plugin name with the binary name.

>     * [ ]   manual install instructions in INSTALL.md mention python instead of python3; also requires plugins/holland.lib.common/ probably in most cases, perhaps this even replaces recommendation of holland.lib.mysql?
> 

Updated the manual install instructions to explicitly call out python3 for install instructions. There is a step to install `holland.lib.common` - it's just worded "Install common plugins". I've updated the wording for that step to be a bit clearer.

>     * [ ]   minor make-config default change for this new provider
> 
> 
> ```
> estimated-size-factor = 1.0
> estimated-size-factor = 0.20
> ```
> 

This actually is a little trickier than I expected. All plugins will currently dump this stanza into a `holland mc`  backupset configuration: https://github.com/holland-backup/holland/blob/master/holland/commands/mk_config.py#L225-L232. However, even if a plugin defines a `[holland:backup]` stanza in the provider.conf - that stanza currently does not influence what gets output by `holland mc`.  We have two potential options:

- Try to have `holland mc` reconcile the `base_config` settings with anything defined in the provider.conf. 
- Just handle it in code and allow a plugin to maybe define a custom `holland:backup` stanza to use when writing a `holland mc ` for a particular plugin. If a custom definition is not defined, then it just goes to dumping the default block of code I noted previously. 
 
>     * [ ]   MariaDB uses PAM auth by default but provider doesn't default to using unix socket but instead uses localhost, I think a standard unix socket path is a good clue to new users rather than intuition of setting a password for existing root db user. Proposed fix, make-config:
> 
> 
> ```
> ### WAS
> [mysql:client]
> defaults-extra-file = ~/.my.cnf,
> # user = "" # no default
> # password = "" # no default
> # socket = "" # no default
> # host = "" # no default
> # port = "" # no default
> ```
> 
> ```
> ### NEW
> [mysql:client]
> defaults-extra-file = ~/.my.cnf,
> user = "root"
> # password = "" # no default, modern mariadb uses pam auth
> socket = "/var/lib/mysql/mysql.sock" # no default
> # host = "" # no default
> # port = "" # no default
> ```

Sure, let me adjust that.

ed-velez commented:
> Looks good, should add a test to the [Jenkinsfile](https://github.com/holland-backup/holland/blob/master/Jenkinsfile)

The only tricky thing with that is having to install mariadb. The jenkinsfile currently is just installing mysql. Any suggestions?

soulen3 commented:
> > Looks good, should add a test to the [Jenkinsfile](https://github.com/holland-backup/holland/blob/master/Jenkinsfile)
> 
> The only tricky thing with that is having to install mariadb. The jenkinsfile currently is just installing mysql. Any suggestions?

Couldn't tell you. I wrote that a while ago
